src/metabase/events/notifications.clj

```diff
@@ -0,0 +1,87 @@
+(ns metabase.events.notifications
+  (:require [clojure.core.async :as async]
+            [clojure.tools.logging :as log]
+            [korma.core :as k]
+            [metabase.config :as config]
+            [metabase.email.messages :as messages]
+            [metabase.events :as events]
+            (metabase.models [card :refer [Card]]
+                             [dashboard :refer [Dashboard]]
+                             [dashboard-card :refer [DashboardCard]]
+                             [dependency :refer [Dependency]]
+                             [pulse :refer [Pulse]]
+                             [pulse-card :refer [PulseCard]]
+                             [user :as user])
+            [metabase.util.urls :as urls]))
+
+
+(def notifications-topics
+  "The `Set` of event topics which are subscribed to for use in notifications tracking."
+  #{:segment-update})
+
+(def ^:private notifications-channel
+  "Channel for receiving event notifications we want to subscribe to for notifications events."
+  (async/chan))
+
+
+;;; ## ---------------------------------------- EVENT PROCESSING ----------------------------------------
+
+
+(defn- send-user-segment-notification [segment updated-by user-id card-deps]
+  (let [dash-deps  (when (seq card-deps)
+                     (k/select Dashboard
+                       (k/join DashboardCard (= :report_dashboardcard.dashboard_id :id))
+                       (k/fields :id :name)
+                       (k/where {:report_dashboardcard.card_id [in (mapv :id card-deps)]})))
+        pulse-deps (when (seq card-deps)
+                     (k/select Pulse
+                       (k/join PulseCard (= :pulse_card.pulse_id :id))
+                       (k/fields :id :name)
+                       (k/where {:pulse_card.card_id [in (mapv :id card-deps)]})))
+        send-to    (user/retrieve-user user-id)]
+    (messages/send-notification-email
+      (:email send-to)
+      (:name segment)
+      (:common_name updated-by)
+      (:email updated-by)
+      (mapv #(assoc % :url (urls/dashboard-url (:id %))) dash-deps)
+      (mapv #(assoc % :url (urls/question-url (:id %))) card-deps)
+      (mapv #(assoc % :url (urls/pulse-url (:id %))) pulse-deps)
+      (:revision_message segment))))
```

Comment:
*  It's almost certainly a bad sign if your function has more than 5 params, let alone 8. I would seriously look into ways to break this function up into smaller, more digestible pieces.
*  At the very least you should probably keywordize the args if you're going to have that many, or it's really going to be hard to keep track of things, and to tweak it in the future
*  You could also abstract out your code that adds URLs to make the function call itself a little less hairy

```clojure
(defn- add-urls-for-ids [id->url deps] ; id->url could not be anything, the sky's the limit! Function, map, etc.
  (for [dep deps]
    (assoc dep :url (id->url (:id dep)))))

(messages/send-notification-email
 email            (:email send-to)
 object-name      (:name segment)
 modifier-name    (:common_name updated-by)
 modifier-email   (:email updated-by)
 dashboards       (add-urls-for-ids urls/dashboard-url dash-deps)
 cards            (add-urls-for-ids urls/pulse-url     pulse-deps)
 pulses           (add-urls-for-ids urls/question-url  card-deps)
 revision-message (:revision_message segment))
```

*  It looks like you're pulling two things out of `updated-by` and two out of `segment`. You could just pass the whole object and pull them out inside `send-notification-email`. Or even make a 6-arity version that pulls them out and passes them to the 8-airity version. If you're gonna have a function with 8 args, probably best to keep the code that calls it as close as humanly possible.